snuba/stateful_consumer/states/bootstrap.py

```diff
@@ -1,25 +1,178 @@
+        """
+        if not self.__active_snapshot_id:
+            return ConsumerStateCompletionEvent.NO_SNAPSHOT
+        elif not self.__loaded_snapshot_transactions:
+            """
+            This means we received an init message but we are still waiting for
+            the snapshot loaded message.
+            """
+            return ConsumerStateCompletionEvent.SNAPSHOT_INIT_RECEIVED
+        else:
+            return ConsumerStateCompletionEvent.SNAPSHOT_READY_RECEIVED
+
+    def get_active_snapshot(self) -> Optional[Tuple[SnapshotId, Optional[TransactionData]]]:
+        if self.__active_snapshot_id:
+            return (self.__active_snapshot_id, self.__loaded_snapshot_transactions)
+        else:
+            return None
+
+    def __commit_if_no_active_snapshot(self) -> CommitDecision:
+        if not self.__active_snapshot_id:
+            return CommitDecision.COMMIT_THIS
+        else:
+            return CommitDecision.DO_NOT_COMMIT
+
+    def process_init(self, msg: SnapshotInit) -> CommitDecision:
+        """
+        Processes a snapshot-init message found on the control topic.
+        If the snapshot-init message is for this consumer and we are not already
+        waiting on an existing snapshot this sets the active snapshot and does not
+        commit the message for failover.
+        """
+        logger.debug("Processing init message for %r", msg.id)
+        if msg.product != settings.SNAPSHOT_LOAD_PRODUCT:
+            return self.__commit_if_no_active_snapshot()
+        if msg.id in self.__past_snapshots:
+            logger.warning(
+                "Duplicate Snapshot init: %r",
+                msg.id,
+            )
+            return self.__commit_if_no_active_snapshot()
+        if self.__active_snapshot_id and not self.__loaded_snapshot_transactions:
+            logger.error(
+                "A new snapshot-init message was found before a previous one "
+                " was completed. Ignoring. Running %r. Init received %r.",
+                msg.id,
+                self.__active_snapshot_id,
+            )
+            return CommitDecision.DO_NOT_COMMIT
+
+        # TODO: filter out the tables I am not interested into.
+
+        self.__past_snapshots.add(msg.id)
+        self.__active_snapshot_id = msg.id
+        self.__loaded_snapshot_transactions = None
+        return CommitDecision.COMMIT_PREV
+
+    def process_abort(self, msg: SnapshotAbort) -> CommitDecision:
+        """
+        Processes an abort message found on the control topic. If this comes after
+        a snapshot-init message, it makes the state machine drop that snapshot.
+        """
+        logger.debug("Processing abort message for %r", msg.id)
+        if msg.id not in self.__past_snapshots:
+            return self.__commit_if_no_active_snapshot()
+        if self.__active_snapshot_id != msg.id:
+            logger.warning(
+                "Aborting a snapshot that is not active. Active %r, Abort %r",
```

Comment:
I think here (and in `process_snapshot_below` it might make sense to reword this as noted above (if you agree with my interpretation of what this is saying.)